src/blog/0.3/generics-hash-codes-madness.vmd

```diff
@@ -0,0 +1,256 @@
+
+```
+$ echo hello
+hello
+$ echo hello
+hello
+$ echo hello
+hello
+$ echo hello
+hello
+$ echo hello
+howdy
+```
+
+
+
+When a program isn't predictable, debugging becomes difficult, birds fall from the sky, and our most precious illusions fade away and leave us in an uncertain, lovecraftian world filled with terrors.
+
+
+This is *non-determinism*. It's when a function or program gives unexpectedly different results even when we give it the same inputs.
+
+
+# The Investigation
+
+
+Every time I ran the compiler, I got a different result. It's like it would randomly decide which data to process first, so I couldn't reliably reproduce the problem I was trying to solve. I spent _hours_ on this, with no luck.
+
+
+"Something's non-deterministic! It's driving me insane!" I said to the server.
+
+
+"But how? Isn't the frontend written in Scala? The JVM is deterministic." someone said.
+
+
+"It should be deterministic," I said, "it's garbage collected after all..."
+
+
+Then I remembered the winter of '19. It was the [7 Day Roguelike Challenge](https://7drl.com/), when developers worldwide spend 7 days each making a roguelike game. It was day five, we'd already lost half of our comrades to bugs, and everyone was in their final days.
+
+
+I had fired up my program for the fifteenth time that day, but things were suddenly acting weird: when I ran the same program with the same inputs, I got different behavior every time. This was a problem, because it was a game where the user could save, reload, and [travel back in time](https://verdagon.itch.io/incendian-falls), expecting things to behave the same way as they did last time. Mastering these techniques was very important for defeating the final boss, a Kraken.
+
+
+After spending most of the day diving and debugging, I'd figured out that it was because I was using strings in a new place, and C#'s `string.GetHashCode()` returns [something different every run](https://andrewlock.net/why-is-string-gethashcode-different-each-time-i-run-my-program-in-net-core/).
+
+
+Needless to say, non-determinism and I have _a history._
+
+
+"But that can't happen in Scala..." I mused to myself, "it always provides a pretty sane default hashCode implementation."
+
+
+"I think there's a couple classes that just return the object's address." someone said.
+
+
+"That's impossible!" I said. "The JVM can't return an address because it literally _moves_ objects around during compaction!"
+
+
+# The Culprit
+
+
+It turns out, they were right. Whereas Scala's `Vector`, `List`, etc. will all calculate their hash codes based on their elements, Scala's `Array` doesn't. It inherits the `hashCode` function from `java.util.Object` which **just returns the address of the object.**
+
+
+And then, to make sure it returns the same hash code for the entire duration of the program, the JVM literally [remembers the object's original address](https://stackoverflow.com/questions/16445094/is-javas-hashcode-deterministic). How crazy is that?
```

Comment:
This link doesn't explain how the JVM keeps track of the object's original address, just that hashCode isn't deterministic.
Also, on stackoverflow you should be linking to a specific answer, not just the whole question.